ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageCompactor.java

```diff
@@ -124,30 +128,30 @@ public int compactPartitions() {
     long compactionStartTime = now;
     long timeToQuit = now + compactionTimeLimitMs;
     long queryEndTime = now - retentionPeriodMs;
-    // TODO: we can cache the latest timestamp that we know we have cleared and use that on subsequent calls
+    // TODO: checkpoint the latest timestamp and use that on subsequent compactions
     // Starting from beginning of time is too expensive
     // Order partitions by earliest time at which dead blob exists
-    // Can start with retention period and go back additional retention periods until no more found
-    long queryStartTime = 1;
+    long queryStartTime = now - TimeUnit.DAYS.toMillis(lookbackDays);
+    Date queryStartDate = new Date(queryStartTime);
+    Date queryEndDate = new Date(queryEndTime);
     int totalBlobsPurged = 0;
     for (PartitionId partitionId : partitionsSnapshot) {
       String partitionPath = partitionId.toPathString();
       if (!partitions.contains(partitionId)) {
         // Looks like partition was reassigned since the loop started, so skip it
         continue;
       }
-      logger.info("Running compaction on partition {}", partitionPath);
+      logger.info("Compacting partition {} over time range {} - {}", partitionPath, queryStartDate, queryEndDate);
       try {
-        // TODO: before compacting, call getOldestBlob to get queryStartTime
         int numPurged =
             compactPartition(partitionPath, CloudBlobMetadata.FIELD_DELETION_TIME, queryStartTime, queryEndTime,
```

Comment:
Can there be a case where we cannot complete compaction of all partitions in 24 hours?  In that case, it's possible that same that partitions that are at the end of the list of partitions, miss the compaction every time. There are atleast 2 ways we can approach this:

1. If instead of draining out each partition in a loop, we did n batches of `compactPartittion` calls for each partition, before looping through the list of partitions again, it will keep every partition compacted uniformly.
2. Another way we can solve this could be by setting cloudConfig.cloudBlobCompactionIntervalHours large enough so that this scenario can never happen. For this, it might be good to have metrics that can tell us how many times we were able to successfully loop through all partitions, and how many times we couldn't loop through all partitions.

Reply:
I think the first suggestion is a good idea.  I will file a separate ticket to do that in a follow up PR.